lib/galaxy/tools/cache.py

```diff
@@ -1,89 +1,89 @@
-        v = json.loads(unicodify(v))
-        if v.get('tool_cache_version', 0) != CURRENT_TOOL_CACHE_VERSION:
-            return NO_VALUE
-        for path, modtime in v['paths_and_modtimes'].items():
+def decoder(obj):
+    return json.loads(zlib.decompress(bytes(obj)).decode('utf-8'))
+
+
+class ToolDocumentCache:
+
+    def __init__(self, cache_dir):
+        self.cache_dir = cache_dir
+        if not os.path.exists(self.cache_dir):
+            os.makedirs(self.cache_dir)
+        self.cache_file = os.path.join(self.cache_dir, 'cache.sqlite')
+        self.writeable_cache_file = None
+        # Create database if necessary using 'c' flag
+        self._cache = SqliteDict(self.cache_file, flag='c', encode=encoder, decode=decoder, autocommit=False)
+        # Switch SqliteDict back to readonly
+        self._cache.flag = 'r'
+
+    def get(self, config_file):
+        tool_document = self._cache.get(config_file)
+        if not tool_document:
+            return None
+        if tool_document.get('tool_cache_version', 0) != CURRENT_TOOL_CACHE_VERSION:
+            return None
+        for path, modtime in tool_document['paths_and_modtimes'].items():
             if os.path.getmtime(path) != modtime:
-                return NO_VALUE
-        payload = get_tool_source(
-            config_file=k,
-            xml_tree=etree.ElementTree(etree.fromstring(v['payload'].encode('utf-8'))),
-            macro_paths=v['macro_paths']
-        )
-        return CachedValue(metadata=v['metadata'], payload=payload)
-
-    def value_encode(self, v):
-        return unicodify(v.payload.to_string())
-
-
-def create_cache_region(tool_cache_data_dir):
-    if not os.path.exists(tool_cache_data_dir):
-        os.makedirs(tool_cache_data_dir)
-    region = make_region()
-    region.configure(
-        'dogpile.cache.dbm',
-        arguments={"filename": os.path.join(tool_cache_data_dir, "cache.dbm")},
-        expiration_time=-1,
-        wrap=[JSONBackend],
-        replace_existing_backend=True,
-    )
-    return region
+                return None
+        return tool_document
+
+    def make_writable(self):
+        if not self.writeable_cache_file:
+            self.writeable_cache_file = RenamedTemporaryFile(self.cache_file)
+            if os.path.exists(self.cache_file):
+                shutil.copy(self.cache_file, self.writeable_cache_file.tmpfile.name)
+            self._cache = SqliteDict(self.writeable_cache_file.tmpfile.name, flag='c', encode=encoder, decode=decoder, autocommit=False)
+
+    def persist(self):
+        if self.writeable_cache_file:
+            self._cache.commit()
+            self.writeable_cache_file.close()
+            self.writeable_cache_file = None
+            self._cache = SqliteDict(self.cache_file, flag='r', encode=encoder, decode=decoder, autocommit=False)
+
+    def set(self, config_file, tool_source):
+        self.make_writable()
```

Comment:
I think it would be better to reopen the cache file inside a `with` statement otherwise other threads/processes would not be able to write to it.

Reply:
So the goal of that with statement is to avoid re-assigning `self._cache` here ?

Reply:
I realise now that's a bad idea since instantiating `SqliteDict` creates a thread to serialise access to the connection.

Why not have it always open for read and write?

Reply:
The idea is the `make_writable` creates a tempfile next to the actual cache where the actual writing happens. When that is done we just the file back into place and re-open in read-only mode (read-only mode is technically not necessary since we call `make_writable` for every action that updates the database). This way multiple processes can write to the database at the same time. If we don't do this we get `database locked` exceptions when you start up multiple separate Galaxy processes at the same time (if the cache database is empty, if it's fully populated we don't ever call any of the update methods) -- we could place a lock around the database file, and serialize access that way, but we haven't had good success with file locking on the Anvil / GCP VMs.